src/tito/tar.py

```diff
@@ -0,0 +1,296 @@
+
+    def calculate_checksum(self, values):
+        """The checksum field is the ASCII representation of the octal value of the simple
+        sum of all bytes in the header block. Each 8-bit byte in the header is added
+        to an unsigned integer, initialized to zero, the precision of which shall be
+        no less than seventeen bits. When calculating the checksum, the checksum field is
+        treated as if it were all spaces.
+
+        Callers of this method are responsible for *not* sending in the previous checksum.
+        """
+
+        new_chksum = 0
+        for val in values:
+            for x in val:
+                new_chksum += ord(x)
+        for blank in " " * 8:
+            new_chksum += ord(blank)
+
+        return "%07o\x00" % new_chksum
+
+    def process_chunk(self, chunk):
+        # Tar archives end with two 512 byte blocks of zeroes
+        if chunk == "\x00" * 512:
+            self.out.write(chunk)
+            self.total_length += len(chunk)
+            if self.last_chunk_was_nulls:
+                self.out.write("\x00" * (self.padded_size(self.total_length, GIT_BLOCK_SIZE) - self.total_length))
+                self.done = True
+            self.last_chunk_was_nulls = True
+            return
+
+        self.last_chunk_was_nulls = False
+
+        chunk_props = self.chunk_to_hash(chunk)
+
+        # This line is the whole purpose of this class!
+        chunk_props['mtime'] = "%011o\x00" % self.timestamp
+
+        # Delete the old checksum since it's now invalid and we don't want to pass
+        # it in to calculate_checksum().
+        del(chunk_props['checksum'])
+        chunk_props['checksum'] = self.calculate_checksum(chunk_props.values())
+
+        # Remove the trailing NUL byte(s) on the end of members
+        for k, v in chunk_props.items():
+            chunk_props[k] = v.rstrip("\x00")
+
+        for member in self.octal_members:
+            # Convert octals to decimal
+            chunk_props[member] = int(chunk_props[member], 8)
+
+        # If there is no global header, we need to create one
+        if self.need_header:
+            # When run against a tree ID, git archive doesn't create
+            # a global header.  The first block is just the header for
+            # the first file.
+            if chunk_props['typeflag'] != 'g':
+                self.create_global_header()
+                self.create_extended_header()
+                self.process_header(chunk_props)
+            else:
+                self.process_header(chunk_props)
+                self.process_extended_header()
+            self.need_header = False
+        else:
+            self.process_header(chunk_props)
+            self.process_file_data(chunk_props['size'])
+
+    def fix(self):
+        try:
+            chunk = self.fh.read(RECORD_SIZE)
+            while chunk != "" and not self.done:
+                self.process_chunk(chunk)
+                chunk = self.fh.read(RECORD_SIZE)
```

Comment:
> This is not correct. read() will read at most RECORD_SIZE, but it can read only part of it - and still not be at EOF. While this is unusual, it can happen. You should put there the second while-loop as in original perl script.

Good point.  Addressed in dc47b4c.